torch/distributed/_shard/sharded_tensor/_ops/math_ops.py

```diff
@@ -4,96 +4,80 @@
-        if len(args) != 2:
-            raise ValueError("Only support binary math op on ShardedTensor for now!")
-        lhs = args[0]
-        rhs = args[1]
-        # Validate types
-        if isinstance(lhs, ShardedTensor) and isinstance(rhs, ShardedTensor):
-            lhs_spec = lhs.sharding_spec()
-            rhs_spec = rhs.sharding_spec()
-            if not isinstance(lhs_spec, ChunkShardingSpec) or not isinstance(rhs_spec, ChunkShardingSpec):
-                raise TypeError("Only ShardedTensor with ChunkShardingSpec supports"
-                                " two ShardedTensor together")
-
-            if lhs.size() == rhs.size() and lhs_spec.dim == rhs_spec.dim:
-                # perform local element-wise math op
-                res = op(lhs.local_tensor(), rhs.local_tensor())
-                return ShardedTensor._init_from_local_tensor(
-                    res,
-                    lhs_spec,
-                    lhs.size(),  # type: ignore[arg-type]
-                    process_group=pg)
-            else:
-                raise RuntimeError("Implicit broadcasting not supported yet!")
+def binary_math_op_impl(op, types, args=(), kwargs=None, pg=None):
+    """
+    Handles ``__torch_function__`` dispatch for the binary math ops
+    such as `torch.add`, `torch.mul`, `torch.div`, etc.
+    This method computes on ShardedTensor, or ShardedTensor op ReplicatedTensor
+    """
+    if len(args) != 2:
+        raise ValueError("Only support binary math op on ShardedTensor for now!")
+    lhs = args[0]
+    rhs = args[1]
+    # Validate types
+    if isinstance(lhs, ReplicatedTensor):
+        assert isinstance(rhs, ShardedTensor)
+        st_size = rhs.size()
+        st_meta = rhs.local_shards()[0].metadata
+        if st_size != lhs.size():
+            # try to broadcast replicated tensor
+            lhs = lhs.expand(st_size)
 
-        elif isinstance(lhs, ReplicatedTensor):
-            assert isinstance(rhs, ShardedTensor)
-            st_size = rhs.size()
-            st_meta = rhs.local_shards()[0].metadata
-            if st_size != lhs.size():
-                # try to broadcast replicated tensor
-                lhs = lhs.expand(st_size)
+        replica_part = narrow_tensor(lhs, st_meta)
+        res = op(replica_part, rhs.local_tensor())
 
-            replica_part = narrow_tensor(lhs, st_meta)
-            res = op(replica_part, rhs.local_tensor())
+        return ShardedTensor._init_from_local_tensor(
+            res,
+            rhs.sharding_spec(),
+            rhs.size(),  # type: ignore[arg-type]
+            process_group=pg)
 
-            return ShardedTensor._init_from_local_tensor(
-                res,
-                rhs.sharding_spec(),
-                rhs.size(),  # type: ignore[arg-type]
-                process_group=pg)
+    elif isinstance(rhs, ReplicatedTensor):
+        assert isinstance(lhs, ShardedTensor)
+        st_size = lhs.size()
+        st_meta = lhs.local_shards()[0].metadata
```

Comment:
(no need to do it in this PR) Do you think for spec agnostic sharded tensor ops, we should consider the case of multiple local shards?

Reply:
Indeed, some ops have very complicated behavior under the circumstances of multiple shards, which we need to define the expected behavior first. I guess for simple ops like element-wise ops, we already implemented in a spec agnostic way.